tests/integration/testlib/vxlan.py

```diff
@@ -0,0 +1,102 @@
+#
+# Copyright (c) 2019 Red Hat, Inc.
+#
+# This file is part of nmstate
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+#
+
+from contextlib import contextmanager
+
+import libnmstate
+
+from libnmstate.schema import Interface
+from libnmstate.schema import InterfaceState
+from libnmstate.schema import VXLAN
+
+
+def vxlans_up(vxlans):
+    vxlans_state = []
+    for vxlan in vxlans:
+        vxlans_state.append(vxlan_up(*vxlan))
+    return {Interface.KEY: vxlans_state}
+
+
+def vxlans_absent(vxlans):
+    vxlans_state = []
+    for vxlan in vxlans:
+        vxlans_state.append(vxlan_absent(*vxlan))
+    return {Interface.KEY: vxlans_state}
+
+
+@contextmanager
+def apply_state(setup_state, teardown_state):
+    try:
+        libnmstate.apply(setup_state)
+        yield setup_state
+    finally:
+        libnmstate.apply(teardown_state, verify_change=False)
+
+
+@contextmanager
+def vxlan_interfaces(vxlans):
+    vxlans_interfaces_absent = []
+    try:
+        setup_state = vxlans_up(vxlans)
+        libnmstate.apply(setup_state)
+        yield setup_state
+    finally:
+
+        # We need a little adaptation to remove remote from the tuple
+        # since it's not needed for absent
+        for vxlan in vxlans:
+            vxlan_id, parent, _ = vxlan
+            vxlans_interfaces_absent.append((vxlan_id, parent))
+
+        libnmstate.apply(
+            vxlans_absent(vxlans_interfaces_absent), verify_change=False
+        )
+
+
+def vxlan_up(vxlan_id, base_if, remote):
+    vxlan_name = vxlan_ifname(vxlan_id, base_if)
+    return {
+        Interface.NAME: vxlan_name,
+        Interface.TYPE: VXLAN.TYPE,
+        Interface.STATE: InterfaceState.UP,
+        VXLAN.CONFIG_SUBTREE: {
+            VXLAN.ID: vxlan_id,
+            VXLAN.BASE_IFACE: base_if,
+            VXLAN.REMOTE: remote,
+            VXLAN.DESTINATION_PORT: 4789,
+        },
+    }
+
+
+def vxlan_absent(vlan_id, base_if):
+    vxlan_name = vxlan_ifname(vlan_id, base_if)
+    return {Interface.NAME: vxlan_name, Interface.STATE: InterfaceState.ABSENT}
+
+
+def vxlan_down(vlan_id, base_if):
+    vxlan_name = vxlan_ifname(vlan_id, base_if)
+    return {
+        Interface.NAME: vxlan_name,
+        Interface.TYPE: VXLAN.TYPE,
+        Interface.STATE: InterfaceState.DOWN,
+    }
+
+
+def vxlan_ifname(vxlan_id, base_if):
```

Comment:
By itself this function works for me, but it seems overused in the above code.
I would expect the name to be passed once on creation and then either re-passed to the relevant functions or parsed from the data itself. Recreating the name make the functions that uses it dependent in this module, we will not be able to reuse/copy it easily (it is more like a smell than a real problem).

Reply:
Ok, let's pass and don't expect the name to always be parent+id